### DISCUSSION TITLE
create.boxplot: how to add multiple lines of math expressions?

### DISCUSSION THREAD

**Discussion by jarbet** [uclahs-cds/public-R-BoutrosLab-plotting-general#130] [Q&A]
For `create.boxplot`, I am trying to add 2 lines of math expressions: a p-value and Spearman correlation (since the group variable is ordinal).

However, when I pass a vector of 2 `expression` objects to `text.labels`, it does not display both elements but instead displays the first element twice:

``` r
library(survival);
library(BoutrosLab.plotting.general);
#> Loading required package: lattice
#> Loading required package: latticeExtra
#> Loading required package: cluster
#> Loading required package: hexbin
#> Loading required package: grid
#> 
#> Attaching package: 'BoutrosLab.plotting.general'
#> The following object is masked from 'package:stats':
#> 
#>     dist

data(cancer);
lung$ph.ecog.factor <- factor(lung$ph.ecog);

# correlation
spearman <- cor.test(
    x = as.numeric(lung$age),
    y = lung$ph.ecog,
    method = 'spearman',
    use = 'complete.obs'
    );
#> Warning in cor.test.default(x = as.numeric(lung$age), y = lung$ph.ecog, : Cannot
#> compute exact p-value with ties
spearman.p <- spearman$p.value;
spearman.rho <- round(spearman$estimate, 2);
names(spearman.rho) <- NULL;
text.pvalue <- display.statistical.result(
    x = spearman.p,
    statistic.type = 'p',
    symbol = ' = '
    );
text.rho <-  as.expression(bquote(rho == .(spearman.rho)));

create.boxplot(
    formula = age ~ ph.ecog.factor,
    data = lung,
    add.stripplot = TRUE,
    add.text = TRUE,
    text.labels = c(text.rho, text.pvalue),
    text.x = c(4, 4),
    text.y = c(max(lung$age, na.rm = TRUE), max(lung$age, na.rm = TRUE) - 3),
    text.fontface = 'plain'
    )
#> Warning in validDetails.text(x): NAs introduced by coercion
```

![](https://i.imgur.com/WMp7zxy.png)

<sup>Created on 2023-06-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by stefaneng** [comment]
You can use `atop` within `bquote`:

```r
text.pvalue <- display.statistical.result(
    x = spearman.p,
    statistic.type = 'p',
    symbol = ' = '
    );

text.label <- as.expression(
  bquote(atop(rho == .(spearman.rho), .(as.character(text.pvalue))))
)

create.boxplot(
    formula = age ~ ph.ecog.factor,
    data = lung,
    add.stripplot = TRUE,
    add.text = TRUE,
    text.labels = text.label,
    text.x = 4,
    text.y = max(lung$age, na.rm = TRUE) - 1,
    text.fontface = 'plain'
    )
```

![image](https://github.com/uclahs-cds/public-R-BoutrosLab-plotting-general/assets/1639487/382c4012-fa5c-4b39-bf95-81e5df34a63b)

**Comment by jarbet** [comment]
> You can use `atop` within `bquote`:

I noticed you used `as.character` for the second expression, which works in this example.  But what if the p-value is really small and you need to use an expression for scientific notation?

For example:

```
text.pvalue <- display.statistical.result(
    x = 0.00000000512,
    statistic.type = 'p',
    symbol = ' = '
    );
```

How could you combine that with `atop` such that the p-value appears like:

![image](https://github.com/uclahs-cds/public-R-BoutrosLab-plotting-general/assets/103967480/1b9eef3a-6dc5-418d-b77e-a2ec5e31e62e)

**Comment by stefaneng** [comment]
> > You can use `atop` within `bquote`:
> 
> I noticed you used `as.character` for the second expression, which works in this example. But what if the p-value is really small and you need to use an expression for scientific notation?
> 
> For example:
> 
> ```
> text.pvalue <- display.statistical.result(
>     x = 0.00000000512,
>     statistic.type = 'p',
>     symbol = ' = '
>     );
> ```
> 
> How could you combine that with `atop` such that the p-value appears like:

Got it:

```r
text.pvalue <- display.statistical.result(
    x = 0.00000000512,
    statistic.type = 'p',
    symbol = ' = '
    );

text.label <- as.expression(
  bquote(atop(rho == .(spearman.rho[[1]]), .(text.pvalue[[1]])))
)

create.boxplot(
    formula = age ~ ph.ecog.factor,
    data = lung,
    add.stripplot = TRUE,
    add.text = TRUE,
    text.labels = text.label,
    text.x = 4,
    text.y = max(lung$age, na.rm = TRUE) - 1,
    text.fontface = 'plain'
    )
```

Source: https://stackoverflow.com/a/59074689

![image](https://github.com/uclahs-cds/public-R-BoutrosLab-plotting-general/assets/1639487/5aa20b13-4290-4286-8d83-8a0bfdbdf84b)

**Comment by pboutros** [comment]
I might be misunderstanding, but take a look at how it's done for `create.km.plot` in BPS, my recollection is that the canonical solution is to create a `legend `and that `text.labels` was explicitly designed for simple cases.  You'll rarely want it for anything complex I suspect.

**Comment by pboutros** [comment]
> > How could you combine that with `atop` such that the p-value appears like:
> 
> Got it:

Yup I think this is entirely avoidable, shouldn't need `atop `or `bquote `for this per se.  I think this is in the category of "could be done, but there are easier ways to do it".  May be worth looking at other published code and verifying they use `legend `rather than `text.labels` as well.

**Comment by jarbet** [comment]
> > > How could you combine that with `atop` such that the p-value appears like:
> > 
> > 
> > Got it:
> 
> Yup I think this is entirely avoidable, shouldn't need `atop `or `bquote `for this per se. I think this is in the category of "could be done, but there are easier ways to do it". May be worth looking at other published code and verifying they use `legend `rather than `text.labels` as well.

I checked `create.km.plot` and it uses a `legend`.  

Another example is `BPG::get.corr.key` which creates an expression object of length 5 that is passed to the `key` argument of `BPG::create.scatterplot`.

I see `create.boxplot` also has a `key` argument, so I should be able to modify the code from `get.corr.key` if I want more than 2 lines.  Otherwise @stefaneng's solution works for 2 lines.

**Comment by jarbet** [comment]
Here's an example using `key` for an `expression` vector of length 3:

``` r
library(survival);
library(BoutrosLab.plotting.general);
#> Loading required package: lattice
#> Loading required package: latticeExtra
#> Loading required package: cluster
#> Loading required package: hexbin
#> Loading required package: grid
#> 
#> Attaching package: 'BoutrosLab.plotting.general'
#> The following object is masked from 'package:stats':
#> 
#>     dist

data(cancer);
lung$ph.ecog.factor <- factor(lung$ph.ecog);

# correlation
spearman <- cor.test(
    x = as.numeric(lung$age),
    y = lung$ph.ecog,
    method = 'spearman',
    use = 'complete.obs'
    );
#> Warning in cor.test.default(x = as.numeric(lung$age), y = lung$ph.ecog, : Cannot
#> compute exact p-value with ties

spearman.p <- spearman$p.value;
spearman.rho <- round(spearman$estimate, 2);
names(spearman.rho) <- NULL;

text.pvalue <- display.statistical.result(
    x = 0.00000000512,
    statistic.type = 'p',
    symbol = ' = '
    );
text.rho <- as.expression(bquote(rho == .(spearman.rho[[1]])));
text.label <- as.expression(
    bquote(atop(rho == .(spearman.rho[[1]]), .(text.pvalue[[1]])))
    );

key.list <- c(text.rho, text.pvalue, expression(gamma))
key <- list(
    text = list(
        lab = key.list, 
        cex = 1
        ), 
    x = 0.7,
    y = 1
    );

create.boxplot(
    formula = age ~ ph.ecog.factor,
    data = lung,
    add.stripplot = TRUE,
    key = key
    )
```

![](https://i.imgur.com/KojfWiZ.png)

<sup>Created on 2023-06-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by stefaneng** [comment]
@jarbet Since there doesn't seem to be an actual issue and you have a good solution now this is probably best as a discussion
